Optional module argument to `@eval` passed to `eval`

StefanKarpinski opened the pull request:
Useful for evaluating definitions in modules, which is especially handy now that #265 is fixed, e.g.:

```jl
julia> @eval Base function f(args...)
    # ...
end
```

Much nicer since you can paste the definition and just hit enter.

StefanKarpinski committed:
Optional module argument to `@eval` passed through to `eval`

Useful for evaluating definitions in modules, which is especially
handy now that #265 is fixed, e.g.:

    julia> @eval Base function f(args...)
        # ...
    end

Much nicer since you can paste the defintion and just hit enter.

StefanKarpinski commented:
I found this quite pleasant. Since it matches the `eval` function it seems like a natural addition.

vtjnash commented:
Seem like a logical step to me, since we can now provide multiple signatures for macros.

vtjnash commented on base/expr.jl:
```diff
@@ -109,13 +109,17 @@ evaluates expressions in that module.
     :($(esc(:eval))($(Expr(:quote,x))))
 end
+macro eval(m, x)
+    :($(esc(:eval))($(esc(m)),$(Expr(:quote,x))))
```
That `:eval` probably shouldn't be in `esc`. Also true of the other macro (it should probably instead splice in the value of `current_module()` as the first argument to `eval`).

StefanKarpinski replied:
Probably true, but I think that's a bugfix rather than a feature addition, so separate PR.

StefanKarpinski replied:
I think this is this way because of a few things:

* Every module has its own `eval` that already evaluates into that module. I believe this can only be avoided by creating a `baremodule`, which is rare, of course.
* Is emitting the code `eval(current_module(), expr)` really equivalent to `eval(expr)`? I used to think I understood `current_module()` but you showed me some behavior that made me confused and think that I don't really understand what it does anymore.

StefanKarpinski committed:
Fix minor hygiene issue in `@eval` macro (closes #20397)

This is Jameson's suggstion of calling `Core.eval` with
`current_module`; there aren't a lot of situations where this
has an effect on code since all modules by default have an
`eval` function that evaluates code inside of them. However,
you can overwrite this binding with a different value, which
would previously cause `@eval` to fail. The test for this also
serves as a test that `@eval` works in general, with and
without the module argument.

StefanKarpinski committed:
Change various uses of `eval` function to use `@eval` macro

Many are clearer and require less punctuation this way.

StefanKarpinski committed:
Merge branch 'master' into sk/eval
